### ISSUE TITLE
`biocmask`

### ISSUE THREAD

**Issue by jtlandis** [Bioconductor/Contributions#3572]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/jtlandis/biocmask

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by hpages** [comment]
Hi @jtlandis,

Thanks for this submission.

This is how a RangedSummarizedExperiment object gets displayed before loading **biocmask**:
```
library(airway)
data(airway
airway
# class: RangedSummarizedExperiment 
# dim: 63677 8 
# metadata(1): ''
# assays(1): counts
# rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492
#   ENSG00000273493
# rowData names(10): gene_id gene_name ... seq_coord_system symbol
# colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
# colData names(9): SampleName cell ... Sample BioSample
```

And this is how it gets displayed after loading **biocmask**:
```
library(biocmask)
airway
# # A RangedSummarizedExperiment-tibble Abstraction: 63,677 × 8
#     .features       .samples  | counts | gene_id gene_name entrezid gene_biotype
#     <chr>           <chr>     |  <int> | <chr>   <chr>        <int> <chr>       
#   1 ENSG00000000003 SRR10395… |    679 | ENSG00… TSPAN6          NA protein_cod…
#   2 ENSG00000000005 SRR10395… |      0 | ENSG00… TNMD            NA protein_cod…
#   3 ENSG00000000419 SRR10395… |    467 | ENSG00… DPM1            NA protein_cod…
#   4 ENSG00000000457 SRR10395… |    260 | ENSG00… SCYL3           NA protein_cod…
#   5 ENSG00000000460 SRR10395… |     60 | ENSG00… C1orf112        NA protein_cod…
#   …        …            …            …      …        …            …       …     
# n-4 ENSG00000273489 SRR10395… |      0 | ENSG00… RP11-180…       NA antisense   
# n-3 ENSG00000273490 SRR10395… |      0 | ENSG00… TSEN34          NA protein_cod…
# n-2 ENSG00000273491 SRR10395… |      0 | ENSG00… RP11-138…       NA lincRNA     
# n-1 ENSG00000273492 SRR10395… |      0 | ENSG00… AP000230…       NA lincRNA     
# n   ENSG00000273493 SRR10395… |      0 | ENSG00… RP11-80H…       NA lincRNA     
## ℹ n = 509,416
## ℹ 16 more variables: gene_seq_start <int>, gene_seq_end <int>,
##   seq_name <chr>, seq_strand <int>, seq_coord_system <int>, symbol <chr>,
##   `` <>, SampleName <fct>, cell <fct>, dex <fct>, albut <fct>, Run <fct>,
##   avgLength <int>, Experiment <fct>, Sample <fct>, BioSample <fct>
```

Loading a package is not supposed to have this kind of side effect. Imagine the confusion that will result in situations where **biocmask** is not explicitely loaded by the user (with `library(biocmask)`), but gets instead loaded as a dependency (direct or indirect) of another package.

I see that this is the consequence of the package redefining the `show()` method for RangedSummarizedExperiment (originally defined in the **SummarizedExperiment** package) with its own. Please do not do that.

More generally speaking, you're not allowed to redefine existing methods with your own. Defining more specialized methods (i.e. defining a `foo()` method for class B when there's already a `foo()` method for parent class A) is ok if done properly. However, redefining an existing method (i.e. defining a `foo()` method for class A when there's already a `foo()` method for class A defined somewhere else) is never ok.

Thanks,
H.

**Comment by mikelove** [comment]
We can certainly take this out. 

We had followed from the behavior in **tidySummarizedExperiment**, but fine to remove this behavior in **biocmask**, as it is not core to the functionality. We could come up with a different function that users can pipe into, at the end of a series of commands, like `showtidy()`.

**Comment by hpages** [comment]
Thanks Mike.

Didn't know that **tidySummarizedExperiment** was also doing this. It should not. So unfortunate! I'll ask them to reconsider.

Maybe I'm overreacting but it's the first time I see something like that in the many years I've been involved with Bioconductor. People are used to see a RangedSummarizedExperiment object displayed a certain way, so I'm glad we agree that _silently_ changing that to present the object as a "RangedSummarizedExperiment-tibble Abstraction" is not ok. This is something that the user needs to explicitely request. Again, it cannot simply be a side effect of loading a package. Sorry for the rant...

So in this case it's probably ok to redefine the `show()` method for SummarizedExperiment objects, but only if the new method does the following:
- by default it just delegates to the original method defined in **SummarizedExperiment**; 
- it uses a global option that the user needs to explicitely set to switch to the "RangedSummarizedExperiment-tibble Abstraction" display.

Concretely this means that global option `restore_SummarizedExperiment_show` would need to be replaced with something like `show_SummarizedExperiment_as_tibble_abstraction` that defaults to `FALSE`.

Also I get this when I load **tidySummarizedExperiment** after **biocmask**:
```
Loading required package: ttservice
Registered S3 methods overwritten by 'tidySummarizedExperiment':
  method                         from    
  filter.SummarizedExperiment    biocmask
  group_by.SummarizedExperiment  biocmask
  mutate.SummarizedExperiment    biocmask
  print.SummarizedExperiment     biocmask
  pull.SummarizedExperiment      biocmask
  select.SummarizedExperiment    biocmask
  summarise.SummarizedExperiment biocmask
  summarize.SummarizedExperiment biocmask
```
That is not good either. What's the relationship between **tidySummarizedExperiment** and **biocmask**? What differentiates the latter from the former? Can't the latter piggy back on the former so it doesn't need to redefine a bunch of things that are already defined in the former? It sounds like communication/coordination/harmonization with the rest of the `tidy*` community in Bioconductor could have been better. 

Finally, about the name of the package: **biocmask**. Given that the package only handles SummarizedExperiment objects and derivatives ("The package provides `rlang` data masks for the SummarizedExperiment class"), wouldn't it make sense to use a more sepcific name? The current name makes it sound like the package provides `rlang` data masks for bioc stuff in general. Maybe you want to choose a name that leaves some room for other packages that will do the same but for other BioC objects?

Thanks,
H.

**Comment by mikelove** [comment]
I think of biocmask as an alternative approach to tidySE -- as they do a subset of common things with different approaches, different efficiency, different convenience, different verbosity. We do need to be more explicit about that in the vignette, and explain when one might want to use one vs the other. But I wouldn't necessarily think we need to have both loaded at once in an R session, because they are alternatives to some degree. I can check with Stefano.

Regarding "bioc" in the name, we were considering extending to other Bioc S4 objects. But maybe they are mostly SE-shaped objects. @jtlandis what do you think?

**Comment by jtlandis** [comment]
Hello @hpages 

Thank you for taking the time to review our submission!

regarding the printing methods - We can change the implementation to which ever you feel best. If we do go with changing the global option to opt-in, is BioConductor okay with the show method being:

```r
setMethod(
  f="show",
  signature="SummarizedExperiment",
  definition=function(object) {
    if (isTRUE(x=getOption(x="show_SummarizedExperiment_as_tibble_abstraction",
                           default = FALSE)) 
    ) {
      print(object)
    } else {
      methods::getMethod(
        f = "show",
        signature = "SummarizedExperiment",
        where=asNamespace(ns = "SummarizedExperiment"))(object)
    }
  }
)
```

or is over-writing this method upon loading still a bad practice? Alternatively we could export two functions:

```r

# to store old method
def_env <- new.env(parent = emptyenv(), size = 2L)

#' @name tidy-show-optin
#' @return method called for side effects
#' @export
use_tidy_show <- function() {
  if (!is.null(def_env$last_def)) {
    stop("`use_tidy_show()` has already been called once!")
  }
  # store the original definition
  # should fail if definition doesn't yet exist
  def_env$last_def <- getMethod(
    f = "show",
    signature = "SummarizedExperiment"
  )
  setMethod(
    f = "show",
    signature = "SummarizedExperiment",
    definition = function(object) {
      tidyShow(object) #defined in biocmask
      invisible(object)
    }
  )
}

#' @export
use_default_show <- function() {
  if (is.null(def_env$last_def)) {
    stop("`use_tidy_show()` has not yet been called!")
  }
  
  removeMethod(
    f = "show",
    signature = "SummarizedExperiment", 
    where = topenv(parent.frame())
  )
  setMethod(
    f = def_env$last_def@generic,
    signature = def_env$last_def@target,
    definition = def_env$last_def@.Data
  )
  def_env$last_def <- NULL
}

```

From the above, which would you prefer?

I also echo @mikelove on the differences between `tidySE` and `biocmask`. At the root, the implementation and syntax is different, forcing users to be more verbose/explicit. At the moment, `biocmask` is only supporting core `dplyr` verbs, whereas `tidySE` is supporting some `tidyr` functions (pivot_longer/pivot_wider/separate/unite)

Regarding the name "biocmask", this package was intended to bring `rlang` masks to Bioconductor S4 objects, hence the name. I plan to eventually release the API that would allow other packages to implement syntax similar to `biocmask` for their structured S4 objects. However, it is still fairly immature and not general enough to be used in other packages (yet!). So for this release cycle, we limited the scope to only the `SummarizedExperiment` class.

**Comment by hpages** [comment]
Hi @jtlandis,

> If we do go with changing the global option to opt-in, is BioConductor okay with the show method being:

Yes, because even if the new method masks the original, by default it forwards to the original.

Thanks for explaining the differences between **tidySummarizedExperiment** and **biocmask**. I realize now that there's some mention of that in the vignette. I still don't think that **biocmask** should mask 8 of the S3 methods defined in **tidySummarizedExperiment**. This will break the former if it got loaded before **biocmask**, and vice-versa. Unless the 8 methods are the same? In this case, could the code duplication be avoided?

Regarding the name "biocmask". It's still not clear to me whether you plan to broaden the scope of the package to other non-SummarizedExperiment derivatives in the future, or whether this will happen in additional packages. If the latter, then choosing "biocmask" for the SummarizedExperiment-specific package maybe is not the best strategy? Anyways, this is a suggestion only. I can live with the current name.

Thanks,
H.

**Comment by jtlandis** [comment]
> Yes, because even if the new method masks the original, by default it forwards to the original.

Okay, I will adopt this change soon. 

> I still don't think that biocmask should mask 8 of the S3 methods defined in tidySummarizedExperiment.

`biocmask` does mask these functions because they are attempting to fulfill similar space but there isn't a good one-to-one mapping in their implementation.  Perhaps we should be more explicit in the README stating that `biocmask` is written as an alternative to `tidySummarizedExperiment` and that loading both will cause conflicts.

There are expressions that `biocmask` can do that `tidySummarizedExperiment` will fail on, and there are expressions in which `tidySummarizedExperiment` will not return a SummarizedExperiment object, but a tibble, but `biocmask` will disallow since a SummarizedExperiment is not recoverable.

> If the latter, then choosing "biocmask" for the SummarizedExperiment-specific package maybe is not the best strategy?

I somewhat agree as I wasn't sure if `biocmask` would be obligated to include dplyr implementations for other S4 objects. I am amendable to changing the name of the package to something more specific like `seplyr` or `plyexperiment` and then splitting the portion of the codebase that generates the masks into `biocmask`. @hpages , is the splitting of the codebase something that could be done this cycle will it have to wait until next cycle?

**Comment by mikelove** [comment]
My two cents is to submit a specialized SE package now (that is, `biocmask` but renamed to indicate that it is SE specific). So we would continue with submission under a new name.

And then next cycle do the rework of code base, potentially moving some of the generic machinary into a package that the SE-specific package loads (if we choose to move in this direction). We will learn a lot from this current SE-specific package being available to users.

**Comment by jtlandis** [comment]
Hello @hpages, 

We have made the following changes

1. `biocmask` does not change the default show method on load or on attach. Users will need to opt in by either setting the global option, or using `use_show_tidy()` or `use_show_default()`.
2. We have removed the S3 print method in favor of a new S3 generic, show_tidy, which will always print `biocmask`'s pretty printing regardless of the global option.
3. Included mentions within the README and Vignette that `biocmask` and `tidySummarizedExperiment` are alternatives to each other. This is supported by adding package events within `biocmask` that check for if `tidySummarizedExperiment` becomes attached or is already attached, and sends a verbose message.

@mikelove and I have chatted and we are okay with changing the name of `biocmask` to better reflect the nature of the package. We are thinking of renaming to `plyxp` where  "ply" is a nod to `dplyr` "xp" represents "experiment". In doing such a renaming, what would be the best way to approach this?
